0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+how a particular domain is ordered when relevant, just as the current language server highlights selected triggers
+for a quantification. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As mentioned in the guide-level explanation, `foreach` loops and sequence comprehensions are both able to
+borrow concepts and implementation substantially from other features. Parsing, resolving, verifying, and compiling
+quantifier domains is already a mature aspect of the Dafny implementation. The most significant implementation burden
+is ensuring that enumeration ordering is deterministic. The existing compilation logic for enumerating a domain
+treats this ordering as pure implementation detail, and applies heuristic optimizations to try to make the search as short
+as possible. Ensuring consistent ordering is an additional constraint on this logic, applied only when the surrounding context
+is a sequence comprehension or `foreach` loop.
+
+Here is a sketch of the rules that define the enumeration ordering of a quantifier domain.
+These rules are chosen as generalizations of the existing semantics for set comprehensions,
+such that building a set from all elements in the ordered enumeration results in the same value.
+
+  * Every boolean expression defines an ordered enumeration of free variable bindings that satisfy it. This enumeration may include duplicate bindings, and the ordering is significant.
+  * The most common base case is an expression of the form `x in C`, where `C` is an expression with a sequence, multiset, or (i)set type. These expressions produce enumerations that bind only `x` to each value in the collection.
+    * For sequences, the ordering is the obvious ordering of values in the sequence.
+    * For (i)sets or multisets, the exact ordering is deterministic but unspecified.
+    * Potentially-infinite sets are only allowed if they are enumerable.
+    * If `x` is not a variable, then the enumeration produces an empty set of bindings once for each time the LHS appears in the RHS collection.
+  * For an expression of the form `A && B`, the enumeration includes all bindings that satisfy both `A` and `B`. The enumeration ordering is defined by ordering first by `A`'s ordering and then by `B`'s, and the multiplicities of bindings is multiplied. See below for the pseudocode that calculates this enumeration.
+    * This means that the `&&` operation is not fully symmetrical, but this is already true as verification considers it to be "short-circuiting": 
+      `b != 0 && a/b > 1` is valid but `a/b > 1 && b != 0` is not.
+    * This definition is chosen such that the multiplicities of results from `A && B` is the same as `B && A`, even if the ordering is not the same.
+  * For an expression of the form `A || B`, the enumeration will simply produce the values enumerated by `A` followed by those enumerated by `B`.
```

Comment:
True, the existing heuristics only consider conjuncts and give up on disjuncts. I may do as you say and trim disjuncts out of scope initially - I just wanted to make sure there was a reasonable answer to the semantics of enumerating disjuncts so we have a firm foundation. If and when there's need for it, though, I'd prefer to just implement it rather than have users rewrite their code.

Reply:
Yup that's true. `a + b` at least strongly implies that a new value needs to be allocated at runtime, whereas enumerating `x in a || x in b` less so. There are definitely ways to optimize that (e.g. the lazy sequence concatenation logic in the C# and Java runtimes) but I like the idea of making it easier to infer the rough runtime cost just by looking at the structure of the code.

Reply:
So if you wanted to enumerate the elements of `a + b` without duplication, you would explicitly write `x in a || (x in b && !x in a)`, correct?

Reply:
Yup that would work! It would only be efficient if both `a` and `b` had efficient `in` tests though.